tools/swift-format/swift-format.cpp

```diff
@@ -0,0 +1,344 @@
+                                                     "lines (both 1-based).\n"
+                                                     "Multiple ranges can be formatted by specifying\n"
+                                                     "several -lines arguments.\n"
+                                                     "Can't be used with -offset and -length.\n"
+                                                     "Can only be used with one input file."),
+                                            cl::cat(SwiftFormatCategory));
+
+    static llvm::cl::opt<bool> UseTabs("usetabs",
+                                       llvm::cl::desc("Use tabs for indentation"),
+                                       cl::cat(SwiftFormatCategory));
+
+    static llvm::cl::opt<unsigned> TabWidth("tabwidth",
+                                            llvm::cl::desc("Assumed width of tab character"),
+                                            cl::cat(SwiftFormatCategory));
+
+    static llvm::cl::opt<unsigned> IndentWidth("indentwidth",
+                                               llvm::cl::desc("Characters to indent"),
+                                               cl::cat(SwiftFormatCategory));
+
+    static cl::opt<bool> Inplace("i",
+                                 cl::desc("Inplace edit <file>s, if specified."),
+                                 cl::cat(SwiftFormatCategory));
+
+    static cl::opt<bool> OutputXML("output-replacements-xml",
+                                   cl::desc("Output replacements as XML."),
+                                   cl::cat(SwiftFormatCategory));
+
+    static cl::list<std::string> FileNames(cl::Positional, cl::desc("[<file> ...]"),
+                                           cl::cat(SwiftFormatCategory));
+
+    CodeFormatOptions FormatOptions;
+
+    class FormatterDocument {
+      SourceManager SM;
+      unsigned BufferID;
+      CompilerInvocation CompInv;
+      std::unique_ptr<ParserUnit> Parser;
+
+      class FormatterDiagConsumer: public swift::DiagnosticConsumer {
+
+        void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                              DiagnosticKind Kind, StringRef Text,
+                              const DiagnosticInfo &Info) override {
+          llvm::errs() << "Parse error: " << Text << "\n";
+        }
+
+      } DiagConsumer;
+
+    public:
+      FormatterDocument(std::unique_ptr<llvm::MemoryBuffer> Buffer) {
+        updateCode(std::move(Buffer));
+      }
+
+      void updateCode( std::unique_ptr<llvm::MemoryBuffer> buff ) {
+        BufferID = SM.addNewSourceBuffer(std::move(buff));
+
+        Parser.reset(
+                     new ParserUnit(SM, BufferID,
+                                    CompInv.getLangOptions(),
+                                    CompInv.getModuleName())
+                     );
+
+        Parser->getDiagnosticEngine().addConsumer(DiagConsumer);
+        auto &P = Parser->getParser();
+
+        bool Done = false;
+        while (!Done) {
```

Comment:
This could probably be written using a `for` loop:

    for (bool Done = false; !Done; Done = P.Tok.is(tok::eof))
      P.parseTopLevel();